zebra-network/src/protocol/external/addr/v2.rs

```diff
@@ -0,0 +1,213 @@
+
+impl From<AddrV2> for Option<MetaAddr> {
+    fn from(addr_v2: AddrV2) -> Self {
+        if let AddrV2::IpAddr {
+            untrusted_last_seen,
+            untrusted_services,
+            ip,
+            port,
+        } = addr_v2
+        {
+            let addr = SocketAddr::new(ip, port);
+
+            Some(MetaAddr::new_gossiped_meta_addr(
+                addr,
+                untrusted_services,
+                untrusted_last_seen,
+            ))
+        } else {
+            None
+        }
+    }
+}
+
+/// Deserialize an `addrv2` entry according to:
+/// https://zips.z.cash/zip-0155#specification
+///
+/// Unimplemented and unrecognised addresses are deserialized as `None`.
+/// Deserialization consumes the bytes for these addresses.
+impl ZcashDeserialize for AddrV2 {
+    fn zcash_deserialize<R: Read>(mut reader: R) -> Result<Self, SerializationError> {
+        // > uint32  Time that this node was last seen as connected to the network.
+        let untrusted_last_seen = (&mut reader).zcash_deserialize_into()?;
+
+        // > Service bits. A CompactSize-encoded bit field that is 64 bits wide.
+        let untrusted_services: CompactSize64 = (&mut reader).zcash_deserialize_into()?;
+        let untrusted_services = PeerServices::from_bits_truncate(untrusted_services.into());
+
+        // > Network identifier. An 8-bit value that specifies which network is addressed.
+        //
+        // See the list of reserved network IDs in ZIP 155.
+        let network_id = reader.read_u8()?;
+
+        // > CompactSize      The length in bytes of addr.
+        // > uint8[sizeAddr]  Network address. The interpretation depends on networkID.
+        let addr: Vec<u8> = (&mut reader).zcash_deserialize_into()?;
+
+        // > uint16  Network port. If not relevant for the network this MUST be 0.
+        let port = reader.read_u16::<BigEndian>()?;
+
+        if addr.len() > MAX_ADDR_V2_ADDR_SIZE {
+            return Err(SerializationError::Parse(
+                "addr field longer than MAX_ADDR_V2_ADDR_SIZE in addrv2 message",
+            ));
+        }
+
+        if network_id == 0x01 {
+            // > 0x01  IPV4  4   IPv4 address (globally routed internet)
+
+            // > Clients MUST reject messages that contain addresses that have
+            // > a different length than specified in this table for a specific network ID,
+            // > as these are meaningless.
+            if addr.len() != ADDR_V2_IPV4_ADDR_SIZE {
+                return Err(SerializationError::Parse(
+                    "IPv4 field length did not match ADDR_V2_IPV4_ADDR_SIZE in addrv2 message",
+                ));
+            }
+
+            // > The IPV4 and IPV6 network IDs use addresses encoded in the usual way
+            // > for binary IPv4 and IPv6 addresses in network byte order (big endian).
+            let ip: [u8; ADDR_V2_IPV4_ADDR_SIZE] = addr.try_into().expect("just checked length");
+            let ip = Ipv4Addr::from(ip);
```

Comment:
Optional suggestion: This could be `IpAddr::from(ip)`, which would remove the need for `ip: ip.into()` in the result, with the downside of a little less explicitness.

One wilder idea would be to have some sort of generic helper function that could be reused for both IPv4 and IPv6:

```rust
fn convert_addr_bytes_to_ip_addr<const SIZE: usize>(
    addr_bytes: Vec<u8>,
) -> Result<IpAddr, SerializationError>
where
    IpAddr: From<[u8; SIZE]>,
{
    // > Clients MUST reject messages that contain addresses that have
    // > a different length than specified in this table for a specific network ID,
    // > as these are meaningless.
    if addr_bytes.len() != SIZE {
        return Err(SerializationError::Parse(
            "IPv4 field length did not match the expected IP address size in addrv2 message",
        ));
    }

    // > The IPV4 and IPV6 network IDs use addresses encoded in the usual way
    // > for binary IPv4 and IPv6 addresses in network byte order (big endian).
    let ip: [u8; SIZE] = addr_bytes.try_into().expect("just checked length");

    Ok(IpAddr::from(ip))
}
```

Reply:
I think this change would be risky, because if Rust added extra `From` impls, Zebra would automatically accept different formats.

I don't think that's particularly likely, but I don't want to have to watch out for it, either.

Reply:
I'm not sure I agree. This would be a private helper function, so Zebra wouldn't automatically accept different formats unless code was added to use it with different sizes. So the effort would be the same, or less considering that we don't have to check or review very similar code blocks to see if they are correct, we only need to see if the helper function is called correctly.

But this is optional, we can discuss this again in the future :slightly_smiling_face:

Reply:
Now I understand how you wanted the new method to be used!

See 50ba6526d.